From c7ca20a6f35dfc6c6f1fcc7407dbf7a4d3e00596 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Fri, 13 Jun 2014 16:29:09 +0100
Subject: [PATCH 229/351] ASoC: florida: Power DMIC ref for active inputs

On the Florida devices the DMIC ref does more than just provide a
reference for the DMIC, so much be powered whenever the input is in use.

Change-Id: I097d0c467596557ac9471734f2bc6985e1cd49f3
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 sound/soc/codecs/arizona.c |   42 ++++++++++++++++++++++++++++++++++++++++++
 sound/soc/codecs/arizona.h |    1 +
 sound/soc/codecs/florida.c |    1 +
 3 files changed, 44 insertions(+)

diff --git a/sound/soc/codecs/arizona.c b/sound/soc/codecs/arizona.c
index 3f2b89e..a41e725 100644
--- a/sound/soc/codecs/arizona.c
+++ b/sound/soc/codecs/arizona.c
@@ -236,6 +236,48 @@ int arizona_init_mono(struct snd_soc_codec *codec)
 }
 EXPORT_SYMBOL_GPL(arizona_init_mono);
 
+static const char * const arizona_dmic_refs[] = {
+	"MICVDD",
+	"MICBIAS1",
+	"MICBIAS2",
+	"MICBIAS3",
+};
+
+static const char * const arizona_dmic_inputs[] = {
+	"IN1L",
+	"IN1R",
+	"IN2L",
+	"IN2R",
+	"IN3L",
+	"IN3R",
+	"IN4L",
+	"IN4R",
+};
+
+int arizona_init_input(struct snd_soc_codec *codec)
+{
+	struct arizona_priv *priv = snd_soc_codec_get_drvdata(codec);
+	struct arizona *arizona = priv->arizona;
+	struct arizona_pdata *pdata = &arizona->pdata;
+	int i, ret;
+	struct snd_soc_dapm_route routes[2];
+
+	memset(&routes, 0, sizeof(routes));
+
+	for (i = 0; i < ARRAY_SIZE(pdata->dmic_ref); ++i) {
+		routes[0].source = arizona_dmic_refs[pdata->dmic_ref[i]];
+		routes[1].source = arizona_dmic_refs[pdata->dmic_ref[i]];
+
+		routes[0].sink = arizona_dmic_inputs[i * 2];
+		routes[1].sink = arizona_dmic_inputs[(i * 2) + 1];
+
+		ret = snd_soc_dapm_add_routes(&codec->dapm, routes, 2);
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(arizona_init_input);
+
 int arizona_init_gpio(struct snd_soc_codec *codec)
 {
 	struct arizona_priv *priv = snd_soc_codec_get_drvdata(codec);
diff --git a/sound/soc/codecs/arizona.h b/sound/soc/codecs/arizona.h
index b17f3d6..3db2264 100644
--- a/sound/soc/codecs/arizona.h
+++ b/sound/soc/codecs/arizona.h
@@ -254,6 +254,7 @@ extern int arizona_set_fll(struct arizona_fll *fll, int source,
 extern int arizona_init_spk(struct snd_soc_codec *codec);
 extern int arizona_init_gpio(struct snd_soc_codec *codec);
 extern int arizona_init_mono(struct snd_soc_codec *codec);
+extern int arizona_init_input(struct snd_soc_codec *codec);
 
 extern int arizona_init_dai(struct arizona_priv *priv, int dai);
 
diff --git a/sound/soc/codecs/florida.c b/sound/soc/codecs/florida.c
index b301dff..0e1c3a0 100644
--- a/sound/soc/codecs/florida.c
+++ b/sound/soc/codecs/florida.c
@@ -2017,6 +2017,7 @@ static int florida_codec_probe(struct snd_soc_codec *codec)
 	arizona_init_spk(codec);
 	arizona_init_gpio(codec);
 	arizona_init_mono(codec);
+	arizona_init_input(codec);
 
 	ret = snd_soc_add_codec_controls(codec, wm_adsp2_fw_controls, 8);
 	if (ret != 0)
-- 
1.7.9.5

